src/build/cargo.rs

```diff
@@ -410,12 +410,18 @@ impl Executor for RlsExecutor {
         {
             let config = self.config.lock().unwrap();
             let crate_type = parse_arg(cargo_args, "--crate-type");
-            // Because we only try to emulate `cargo test` using `cargo check`, so for now
-            // assume crate_type arg (i.e. in `cargo test` it isn't specified for --test targets)
-            // and build test harness only for final crate type
-            let crate_type = crate_type.expect("no crate-type in rustc command line");
-            let build_lib = *config.build_lib.as_ref();
-            let is_final_crate_type = crate_type == "bin" || (crate_type == "lib" && build_lib);
+
+            let is_final_crate_type = if config.all_targets {
+                // on all_targets is true all rustc args are fine - skip extra processing
+                false
```

Comment:
I don't understand why this would always be false if we are building all targets - I would have thought that if we're building the bin, even for testing, it would still be the 'final crate'